### PR TITLE
My pass at a more strict eslint profile

### PR DESCRIPTION
**Pull request by jcwilson**
Notable changes called out in the comments

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -74,26 +74,26 @@
     "no-dupe-args": "error",              // disallow duplicate arguments in function definitions
     "no-dupe-keys": "error",              // disallow duplicate keys in object literals
     "no-duplicate-case": "error",         // disallow duplicate case labels
-    "no-empty": 0,                        // disallow empty block statements
+    "no-empty": "error",                  // disallow empty block statements
```
Comment (jcwilson) [inline .eslintrc:77]: this doesn't prevent empty function definitions or empty object definitions - which is good

File: .eslintrc [hunk 2]
```diff
@@ -74,26 +74,26 @@
...
     "no-empty-character-class": 0,        // disallow empty character classes in regular expressions
-    "no-ex-assign": 0,                    // disallow reassigning exceptions in catch clauses
+    "no-ex-assign": "error",              // disallow reassigning exceptions in catch clauses
     "no-extra-boolean-cast": "error",     // disallow unnecessary boolean casts
     "no-extra-parens": 0,                 // disallow unnecessary parentheses
     "no-extra-semi": "error",             // disallow unnecessary semicolons
-    "no-func-assign": 0,                  // disallow reassigning function declarations
-    "no-inner-declarations": 0,           // disallow variable or function declarations in nested blocks
-    "no-invalid-regexp": 0,               // disallow invalid regular expression strings in RegExp constructors
+    "no-func-assign": "error",            // disallow reassigning function declarations
```
Comment (jcwilson) [inline .eslintrc:83]: helps prevent accidental name stomping

File: .eslintrc [hunk 3]
```diff
@@ -74,26 +74,26 @@
...
+    "no-inner-declarations": "error",     // disallow variable or function declarations in nested blocks
```
Comment (jcwilson) [inline .eslintrc:84]: not sure why this would be any different than `vars-on-top`, but `¯\_(ツ)_/¯`

File: .eslintrc [hunk 4]
```diff
@@ -74,26 +74,26 @@
...
     "no-sparse-arrays": 0,                // disallow sparse arrays
     "no-template-curly-in-string": 0,     // disallow template literal placeholder syntax in regular strings
-    "no-unexpected-multiline": 0,         // disallow confusing multiline expressions
+    "no-unexpected-multiline": "error",   // disallow confusing multiline expressions
     "no-unreachable": "error",            // disallow unreachable code after return, throw, continue, and break statements
-    "no-unsafe-finally": 0,               // disallow control flow statements in finally blocks
-    "no-unsafe-negation": 0,              // disallow negating the left operand of relational operators
-    "use-isnan": "error",                 // require calls to isNaN() when checking for NaN
+    "no-unsafe-finally": "error",         // disallow control flow statements in finally blocks
+    "no-unsafe-negation": "error",        // disallow negating the left operand of relational operators
```
Comment (jcwilson) [inline .eslintrc]: prevents what are most likely typos

File: .eslintrc [hunk 5]
```diff
@@ -74,26 +74,26 @@
...
+    "use-isnan": 0,                       // require calls to isNaN() when checking for NaN
```
Comment (jcwilson) [inline .eslintrc]: hmm, going to revert this. not sure why it got set to 0

File: .eslintrc [hunk 6]
```diff
@@ -105,23 +105,24 @@
 
     "accessor-pairs": 0,                  // enforce getter and setter pairs in objects
     "array-callback-return": 0,           // enforce return statements in callbacks of array methods
-    "block-scoped-var": 0,                // enforce the use of variables within the scope they are defined
+    "block-scoped-var": "error",          // enforce the use of variables within the scope they are defined
```
Comment (jcwilson) [inline .eslintrc:108]: also related to `vars-on-top`

File: .eslintrc [hunk 7]
```diff
@@ -105,23 +105,24 @@
...
+          "^[a-z]+(_[a-z]+)+$"            // allow snake-case
       }
     ],
-    "eqeqeq": "error",                          // require the use of === and !==
+    "eqeqeq": "error",                    // require the use of === and !==
     "guard-for-in": 0,                    // require for-in loops to include an if statement
-    "no-alert": "error",                  // disallow the use of alert, confirm, and prompt
-    "no-caller": "error",                 // disallow the use of arguments.caller or arguments.callee
+    "no-alert": 0,                        // disallow the use of alert, confirm, and prompt
+    "no-caller": 0,                       // disallow the use of arguments.caller or arguments.callee
```
Comment (jcwilson) [inline .eslintrc]: will revert

File: .eslintrc [hunk 8]
```diff
@@ -132,47 +133,47 @@
     "no-extend-native": 0,                // disallow extending native types
     "no-extra-bind": 0,                   // disallow unnecessary calls to .bind()
     "no-extra-label": 0,                  // disallow unnecessary labels
-    "no-fallthrough": 0,                  // disallow fallthrough of case statements
-    "no-floating-decimal": "error",       // disallow leading or trailing decimal points in numeric literals
+    "no-fallthrough": "error",            // disallow fallthrough of case statements
```
Comment (jcwilson) [inline .eslintrc]: http://archive.oreilly.com/pub/a/javascript/excerpts/javascript-good-parts/bad-parts.html#switch_fall_through

File: .eslintrc [hunk 9]
```diff
@@ -132,47 +133,47 @@
...
+    "no-floating-decimal": 0,             // disallow leading or trailing decimal points in numeric literals
```
Comment (jcwilson) [inline .eslintrc]: will revert

File: .eslintrc [hunk 10]
```diff
@@ -132,47 +133,47 @@
...
+    "no-implied-eval": "error",           // disallow the use of eval()-like methods
+    "no-invalid-this": "error",           // disallow this keywords outside of classes or class-like objects
     "no-iterator": "error",               // disallow the use of the __iterator__ property
     "no-labels": 0,                       // disallow labeled statements
     "no-lone-blocks": 0,                  // disallow unnecessary nested blocks
     "no-loop-func": "error",              // disallow function declarations and expressions inside loop statements
-    "no-magic-numbers": "error",          // disallow magic numbers
-    "no-multi-spaces": 0,                 // disallow multiple spaces
-    "no-multi-str": 0,                    // disallow multiline strings
+    "no-magic-numbers": 0,                // disallow magic numbers
```
Comment (jcwilson) [inline .eslintrc]: will revert

File: .eslintrc [hunk 11]
```diff
@@ -132,47 +133,47 @@
...
+    "no-multi-spaces": "error",           // disallow multiple spaces
```
Comment (jcwilson) [inline .eslintrc]: i've seen this in our code, and it shouldn't be there

File: .eslintrc [hunk 12]
```diff
@@ -132,47 +133,47 @@
...
+    "no-multi-str": "error",              // disallow multiline strings
     "no-new": 0,                          // disallow new operators outside of assignments or comparisons
     "no-new-func": "error",               // disallow new operators with the Function object
     "no-new-wrappers": "error",           // disallow new operators with the String, Number, and Boolean objects
-    "no-octal": "error",                  // disallow octal literals
+    "no-octal": 0,                        // disallow octal literals
```
Comment (jcwilson) [inline .eslintrc]: will revert

File: .eslintrc [hunk 13]
```diff
@@ -132,47 +133,47 @@
...
     "no-octal-escape": 0,                 // disallow octal escape sequences in string literals
     "no-param-reassign": "error",         // disallow reassigning function parameters
     "no-proto": 0,                        // disallow the use of the __proto__ property
-    "no-redeclare": 0,                    // disallow variable redeclaration
+    "no-redeclare": "error",              // disallow variable redeclaration
```
Comment (jcwilson) [inline .eslintrc:157]: name stomping again

File: .eslintrc [hunk 14]
```diff
@@ -132,47 +133,47 @@
...
     "no-restricted-properties": [         // disallow certain properties on certain objects
       "error", {
         "object": "Math",
         "property": "pow",
         "message": "Use the exponentiation operator (**) instead.."
       }
     ],
-    "no-return-assign": 0,                // disallow assignment operators in return statements
+    "no-return-assign": "error",          // disallow assignment operators in return statements
```
Comment (jcwilson) [inline .eslintrc:165]: this just makes sense

File: .eslintrc [hunk 15]
```diff
@@ -132,47 +133,47 @@
...
     "no-return-await": 0,                 // disallow unnecessary return await
     "no-script-url": 0,                   // disallow javascript: urls
     "no-self-assign": "error",            // disallow assignments where both sides are exactly the same
     "no-self-compare": "error",           // disallow comparisons where both sides are exactly the same
     "no-sequences": 0,                    // disallow comma operators
-    "no-throw-literal": 0,                // disallow throwing literals as exceptions
+    "no-throw-literal": "error",          // disallow throwing literals as exceptions
     "no-unmodified-loop-condition": 0,    // disallow unmodified loop conditions
-    "no-unused-expressions": "error",     // disallow unused expressions
+    "no-unused-expressions": 0,           // disallow unused expressions
```
Comment (jcwilson) [inline .eslintrc]: will revert

File: .eslintrc [hunk 16]
```diff
@@ -181,9 +182,9 @@
     "prefer-promise-reject-errors": 0,    // require using Error objects as Promise rejection reasons
     "radix": "error",                     // enforce the consistent use of the radix argument when using parseInt()
     "require-await": 0,                   // disallow async functions which have no await expression
-    "vars-on-top": 0,                     // require var declarations be placed at the top of their containing scope
+    "vars-on-top": "error",               // require var declarations be placed at the top of their containing scope
     "wrap-iife": "error",                 // require parentheses around immediate function invocations
-    "yoda": 0,                            // require or disallow “Yoda” conditions
+    "yoda": "error",                      // require or disallow “Yoda” conditions
```
Comment (jcwilson) [inline .eslintrc:187]: i could relent on this without much argument, but our style checker should already complain if we're doing an assignment in a conditional expression. Besides, how did you manage to mistake `===` for `=`?

File: .eslintrc [hunk 17]
```diff
@@ -211,60 +212,55 @@
...
+    "func-style": "error",                 // enforce the consistent use of either function declarations or expressions
     "function-paren-newline": 0,           // enforce consistent line breaks inside function parentheses
     "id-blacklist": 0,                     // disallow specified identifiers
     "id-length": 0,                        // enforce minimum and maximum identifier lengths
     "id-match": 0,                         // require identifiers to match a specified regular expression
     "indent": ["error", 4, {               // enforce consistent indentation
       "ignoredNodes": [
-        ":nth-child(1)"
+        ":nth-child(1)",
+        ":nth-child(2)"
```
Comment (jcwilson) [inline .eslintrc:237]: this allows us to deal with the the requirejs/angular boilerplate in a more palatable fashion

File: .eslintrc [hunk 18]
```diff
@@ -211,60 +212,55 @@
...
     }],
     "jsx-quotes": 0,                       // enforce the consistent use of either double or single quotes in JSX attributes
-    "key-spacing": 0,                      // enforce consistent spacing between keys and values in object literal properties
-    "keyword-spacing": [                   // enforce consistent spacing before and after keywords
-      "error", {
+    "key-spacing": ["error", {             // enforce consistent spacing between keys and values in object literal properties
+        "beforeColon": false,
+        "afterColon": true,
+        "mode": "strict"
+    }],
```
Comment (jcwilson) [inline .eslintrc:245]: if anyone has strongly held opinions here, let's hear them

File: .eslintrc [hunk 19]
```diff
@@ -290,24 +286,46 @@
...
     "no-whitespace-before-property":       // disallow whitespace before properties
       "error",
     "nonblock-statement-body-position": 0, // enforce the location of single-line statements
-    "object-curly-newline": 0,             // enforce consistent line breaks inside braces
+    "object-curly-newline": [              // enforce consistent line breaks inside braces
+      "error",
+      {
+        "minProperties": 2
+      }
+    ],
```
Comment (jcwilson) [inline .eslintrc]: allows `foo({ bar: 4 })` but not `foo({ bar: 4, day: 8 })`

**Review by rcorell — APPROVED** [review]

**Pull request merged**